XForm/XForm/Commands/CosmosUnescape.cs

```diff
@@ -0,0 +1,145 @@
+                    tempMap.Add(String8.Convert("#NULL#", new byte[6]), String8.Empty);
+                    tempMap.Add(String8.Convert("#HASH#", new byte[6]), String8.Convert("#", new byte[1]));
+
+                    _replacementMap = tempMap;
+                }
+
+                return _replacementMap;
+            }
+        }
+
+        public override Func<DataBatch> ColumnGetter(int columnIndex)
+        {
+            String8[] transformedArray = null;
+            String8Block block = new String8Block();
+
+            // Pass through columns other than the one being converted
+            if (!_sourceColumns.Contains(columnIndex)) return _source.ColumnGetter(columnIndex);
+
+            // Cache the function to get the source data
+            Func<DataBatch> sourceGetter = _source.ColumnGetter(columnIndex);
+
+            return () => Unescape(sourceGetter(), block, ref transformedArray);
+        }
+
+        private DataBatch Unescape(DataBatch batch, String8Block block, ref String8[] transformedArray)
+        {
+            block.Clear();
+            Allocator.AllocateToSize(ref transformedArray, batch.Count);
+
+            String8[] sourceArray = (String8[])batch.Array;
+            for (int i = 0; i < batch.Count; ++i)
+            {
+                String8 originalValue = sourceArray[batch.Index(i)];
+                transformedArray[i] = this.Unescape(originalValue, block);
+            }
+
+            return DataBatch.All(transformedArray, batch.Count);
+        }
+
+        public String8 Unescape(String8 originalValue, String8Block block)
+        {
+            int startIndex;
+            int endIndex = -1;
+            int currentIndex = 0;
+            String8 newValue = String8.Empty;
+
+            while (true)
+            {
+                // Find the start of the escaped string
+                startIndex = originalValue.IndexOf(_bookmarkChar, currentIndex);
+
+                if (startIndex >= 0)
+                {
+                    // Find the end of the escaped string
+                    endIndex = originalValue.IndexOf(_bookmarkChar, startIndex + 1);
+                }
+
+                // If there's no escaped string
+                if (startIndex < 0 || endIndex < 0)
+                {
+                    break;
+                }
+
+                // Found a potential escaped string
+                String8 escapedValue = originalValue.Substring(startIndex, endIndex - startIndex + 1);
+                String8 unescapedValue;
+
+                if (!this.ReplacementMap.TryGetValue(escapedValue, out unescapedValue))
+                {
+                    // Since the escaped value wasn't one of the known escaped values, do not replace with the unescaped value.
+                    unescapedValue = originalValue.Substring(startIndex, endIndex - startIndex);
+                    endIndex = endIndex - 1;
```

Comment:
It looks like this means replace will resume right at the second '#', so it'll be re-evaluated as the start of an escaped symbol, right? Worth checking I guess that ##N#, for example, unescapes the #N# so you know it works with the shortest gap between a non-symbol and a symbol.